api/src/service/subproject_cache_helper.ts

```diff
@@ -0,0 +1,43 @@
+import { Ctx } from "../lib/ctx";
+import * as Result from "../result";
+import { ConnToken } from "./conn";
+import * as Subproject from "./domain/workflow/subproject";
+import * as Project from "./domain/workflow/project";
+import * as SnapshotService from "./cache_snapshot";
+import { Item } from "./liststreamitems";
+
+export async function getSubproject(
+  conn: ConnToken,
+  ctx: Ctx,
+  projectId: Project.Id,
+  subprojectId: Subproject.Id,
+): Promise<Result.Type<Subproject.Subproject>> {
+  return await SnapshotService.getLatestSnapshot(
+    ctx,
+    conn,
+    projectId,
+    subprojectId,
+    "subproject_snapshot_published",
+  );
+}
+
+export async function getAllSubprojects(
+  conn: ConnToken,
+  ctx: Ctx,
+  projectId: Project.Id,
+): Promise<Subproject.Subproject[]> {
+  const rpcClient = conn.multichainClient.getRpcClient();
+  let items: Item[] = [];
+  items = await rpcClient.invoke("liststreamkeyitems", projectId, "subprojects", false, 0x7fffffff);
+  let subprojects: Array<Subproject.Subproject> = [];
+  let i;
+  for (i = 0; i <= items.length - 1; i++) {
```

Comment:
same as above

Reply:
For-loop is basic and makes the intend clear and code more readable. Also I might be wrong but as far as I know Array native functions like map and reduce are slower in comparison to naive for-loop approach, so that's why i think it would be overkill to use reduce for such simple example.